### PR TITLE
Graham/fh 496 nix installer incorrectly validates fstab entries

### PR DESCRIPTION
**Pull request by grahamc**
##### Description

Drastically simplify how we edit the fstab. Also, enable nosuid and noatime, since neither of those are necessary for /nix.

##### Checklist

- [ ] Formatted with `cargo fmt`
- [ ] Built with `nix build`
- [ ] Ran flake checks with `nix flake check`
- [ ] Added or updated relevant tests (leave unchecked if not applicable)
- [ ] Added or updated relevant documentation (leave unchecked if not applicable)
- [ ] Linked to related issues (leave unchecked if not applicable)

##### Validating with `install.determinate.systems`

If a maintainer has added the `upload to s3` label to this PR, it will become available for installation via `install.determinate.systems`:

```shell
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/pr/$PR_NUMBER | sh -s -- install
```

### CODE REVIEW

File: src/action/macos/create_fstab_entry.rs [hunk 1]
```diff
@@ -144,64 +84,39 @@ impl Action for CreateFstabEntry {
...
-                        *line = fstab_entry(&uuid);
-                        updated_line = true;
-                        break;
-                    }
+        let mut line_present = false;
+        let mut current_fstab_lines = fstab_buf
+            .lines()
+            .filter_map(|line| {
+                // Delete nix-installer entries with a "prelude" comment
+                if line.starts_with("# nix-installer") {
```
Comment (cole-h) [inline src/action/macos/create_fstab_entry.rs:98]: Let's use more of the old prelude comment
```suggestion
                if line.starts_with("# nix-installer created volume labelled") {
```

File: src/action/macos/create_fstab_entry.rs [hunk 2]
```diff
@@ -236,76 +148,67 @@ impl Action for CreateFstabEntry {
...
-    prelude_comment + "\n" + &fstab_entry
-}
-
-fn fstab_prelude_comment(apfs_volume_label: &str) -> String {
-    format!("# nix-installer created volume labelled `{apfs_volume_label}`")
-}
-
 fn fstab_entry(uuid: &Uuid) -> String {
-    format!("UUID={uuid} /nix apfs rw,noauto,nobrowse,suid,owners")
+    format!("UUID={uuid} /nix apfs rw,noatime,noauto,nobrowse,nosuid,owners # Added by the Determinate Nix Installer")
```
Comment (cole-h) [inline src/action/macos/create_fstab_entry.rs:204]: This now differs a lot more from how upstream configures their volume:

https://github.com/NixOS/nix/blob/a8a572b11b3f29cdb2f57a94e32cf72a5ecbf6dc/scripts/create-darwin-volume.sh#L466

These should go in a separate commit with rationale behind the change (other than "I wanted to"....), please.

File: src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#![allow(clippy::unnecessary_filter_map)]
```
Comment (cole-h) [inline src/lib.rs:1]: Let's comment why we added this, please.

File: src/action/macos/create_fstab_entry.rs (hunk 2 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs:204]: https://github.com/DeterminateSystems/nix-installer/pull/1338/commits/7bb01781cb7b676e286883f5b56fc1e22f91b054 I think I did that?

File: src/lib.rs (hunk 3 above)
Comment (grahamc) [reply src/lib.rs:2]: Done!

File: src/action/macos/create_fstab_entry.rs (hunk 1 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs:98]: Done!

File: src/action/macos/create_fstab_entry.rs [hunk 4]
```diff
@@ -236,76 +148,67 @@ impl Action for CreateFstabEntry {
...
+            .filter_map(|line| {
+                // Delete nix-installer entries with a "prelude" comment
+                if line.starts_with("# nix-installer created volume labelled") {
+                    None
+                } else {
+                    Some(line)
+                }
+            })
+            .filter_map(|line| {
+                if line.split(&[' ', '\t']).nth(2) == Some("/nix") {
```
Comment (cole-h) [inline src/action/macos/create_fstab_entry.rs:176]: I think this has always been wrong, which may have caused our issue..... :|

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0a52701d536dacd5ae262681fd602647

This is looking for the "type" (i.e. apfs, ext4, ....) to match `/nix` lol

File: src/action/macos/create_fstab_entry.rs (hunk 2 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs:204]: Updated the commit text!

File: src/action/macos/create_fstab_entry.rs [hunk 5]
```diff
@@ -236,76 +148,67 @@ impl Action for CreateFstabEntry {
...
+            .await
+            .map_err(|e| Self::error(ActionErrorKind::Seek(fstab_path.to_owned(), e)))?;
+        fstab
+            .set_len(0)
+            .await
+            .map_err(|e| Self::error(ActionErrorKind::Truncate(fstab_path.to_owned(), e)))?;
+        fstab
+            .write_all(updated_buf.as_bytes())
+            .await
+            .map_err(|e| Self::error(ActionErrorKind::Write(fstab_path.to_owned(), e)))?;
```
Comment (cole-h) [inline src/action/macos/create_fstab_entry.rs:197]: I wonder if we should replace this with our `write-tmp-then-move` approach while we're here.

File: src/action/macos/create_fstab_entry.rs (hunk 1 above)
Comment (cole-h) [reply src/action/macos/create_fstab_entry.rs:98]: Looks like you did it and then un-did it? :laughing:

File: src/action/macos/create_fstab_entry.rs (hunk 4 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs]: Fixed!

File: src/action/macos/create_fstab_entry.rs [hunk 6]
```diff
@@ -236,76 +148,67 @@ impl Action for CreateFstabEntry {
...
-    prelude_comment + "\n" + &fstab_entry
-}
-
-fn fstab_prelude_comment(apfs_volume_label: &str) -> String {
-    format!("# nix-installer created volume labelled `{apfs_volume_label}`")
-}
-
 fn fstab_entry(uuid: &Uuid) -> String {
-    format!("UUID={uuid} /nix apfs rw,noauto,nobrowse,suid,owners")
+    format!("UUID={uuid} /nix apfs rw,noatime,noauto,nobrowse,nosuid,owners # Added by the Determinate Nix Installer")
```
Comment (cole-h) [inline src/action/macos/create_fstab_entry.rs:204]: We should probably send this upstream (sans the comment of course lol) as well.

File: src/action/macos/create_fstab_entry.rs (hunk 1 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs]: d'oh

File: src/action/macos/create_fstab_entry.rs (hunk 5 above)
Comment (grahamc) [reply src/action/macos/create_fstab_entry.rs:197]: Good idea, done!

**Review by cole-h — APPROVED** [review]

File: src/action/macos/create_fstab_entry.rs (hunk 6 above)
Comment (cole-h) [reply src/action/macos/create_fstab_entry.rs:179]: https://github.com/NixOS/nix/pull/12016
